applications/StructuralMechanicsApplication/custom_strategies/custom_schemes/explicit_central_differences_scheme.hpp

```diff
@@ -0,0 +1,900 @@
+            if (i->HasDofFor(ROTATION_Z))
+            {
+              DoF = 3;
+              fix_rotation[1] = (i->pGetDof(ROTATION_Z))->IsFixed();  
+            }
+
+            for (int j = 0; j < DoF; j++)
+            {
+              if (fix_rotation[j])
+              {
+                r_current_angular_acceleration[j] = 0.00;
+                r_middle_angular_velocity[j] = 0.00;
+              }
+               
+              r_middle_angular_velocity[j]   = 0.00 + (mTime.Middle - mTime.Previous) * r_current_angular_acceleration[j] ; 
+              r_current_angular_velocity[j]  = r_middle_angular_velocity[j] + (mTime.Previous - mTime.PreviousMiddle) * r_current_angular_acceleration[j];
+              //current_rotation[j]          = 0.00;
+            }//for DoF
+          }// Rot DoF
+
+
+
+
+        }//for node
+
+      }//parallel
+
+      mTime.Previous = mTime.Current;
+      mTime.PreviousMiddle = mTime.Middle;
+      KRATOS_CATCH("")
+    }
+
+
+  //***************************************************************************
+  //***************************************************************************
+  void Calculate_RHS_Contribution(Element::Pointer pCurrentElement,
+          LocalSystemVectorType& RHS_Contribution,
+          Element::EquationIdVectorType& EquationId,
+          ProcessInfo& rCurrentProcessInfo) 
+  {
+
+    KRATOS_TRY
+
+    //basic operations for the element considered
+    Matrix dummy_lhs;
+    pCurrentElement -> CalculateLocalSystem(dummy_lhs,RHS_Contribution,rCurrentProcessInfo);
+
+    //add explicit contribution of the Element Residual (RHS) to nodal Force Residual (nodal RHS)
+    (pCurrentElement) -> AddExplicitContribution(RHS_Contribution, RESIDUAL_VECTOR, FORCE_RESIDUAL, rCurrentProcessInfo);
+    if (pCurrentElement->GetGeometry()[0].HasDofFor(ROTATION_X))
+    {
+      (pCurrentElement) -> AddExplicitContribution(RHS_Contribution, RESIDUAL_VECTOR, MOMENT_RESIDUAL, rCurrentProcessInfo);
+    }
+    
+    KRATOS_CATCH( "" )
+  }
+
+  //Elements:
+  //****************************************************************************
+
+  void GetFirstDerivativesVector(Element::Pointer pCurrentElement, Vector& rValues ) //V at time n-1/2 old
+  {
+
+    const unsigned int number_of_nodes = pCurrentElement->GetGeometry().size();
+    const unsigned int dimension       = 3;
+    unsigned int       element_size    = number_of_nodes * dimension;
+    const bool check_has_rot_dof = pCurrentElement->GetGeometry()[0].HasDofFor(ROTATION_X);
+    
+    if (check_has_rot_dof) element_size *= 2;
+
+    if ( rValues.size() != element_size ) rValues.resize( element_size, false );
+
+    for ( unsigned int i = 0; i < number_of_nodes; i++ )
+      {
+        unsigned int index = i * dimension;
+        if (check_has_rot_dof) index *= 2;
+
+
+        pCurrentElement->GetGeometry()[i].GetValue(MIDDLE_VELOCITY);
+
+        rValues[index]     = pCurrentElement->GetGeometry()[i].GetValue( MIDDLE_VELOCITY )[0];
```

Comment:
access MIDDLE_VELOCITY onces as a vector, and then to each components. this way you will minimize accesses to the array

Reply:
this function was useless, sry for that